### ISSUE TITLE
How to enforce the optimized motion to be consistent with the terrain

### ISSUE THREAD

**Issue by orsoromeo** [ethz-adrl/towr#6]
First of all congratulation for the useful toolbox!
I have been playing with the towr user interface trying different gaits, terrains and base pose targets. The solver seems to ignore the collision between the feet and the terrain, with the optimized trajectories eventually causing the feet to penetrate into the obstacles. I initially thought it was due to a very coarse default tolerance in IFOPT but, even after decreasing the **tol** to 1e-7 ipopt option, the optimized trajectory does not seem to improve.  How can I enforce the complementarity constraint between terrain and the feet and obtain complex motions similar to those explained in this [paper](https://www.youtube.com/watch?v=0jE46GqzxMM&feature=youtu.be)?

**Comment by awinkler** [comment]
Hi, could you kindly tell me for which terrain this is not working?
It could be there is a mismatch between the rviz visualization [(rviz_terrain_builder)](https://github.com/ethz-adrl/towr/blob/dee989cfde079ad127218c48a184dd559742b149/towr_ros/src/rviz_terrain_builder.cc) and the actual height map the solver is using [(height_map_examples)](https://github.com/ethz-adrl/towr/blob/dee989cfde079ad127218c48a184dd559742b149/towr_ros/src/height_map_examples.cc).

**Comment by orsoromeo** [comment]
This is actually happening for all the terrains, even on the default flat terrain.  For a flat terrain if I set the target base pose to be under the terrain surface the robots will simply go there and penetrate the terrain (see pic). Moreover the feet during the stance phase slide horizontally whilst I expect them to have zero horizontal velocity instead.
![feet_penetration](https://user-images.githubusercontent.com/14978496/39061771-b09385f6-44c5-11e8-9973-713900f630f4.png)

**Comment by awinkler** [comment]
Thanks, I'll look into this. Can you make sure you are on the master branch and toggling different terrains with "t" through the user interface.

**Comment by orsoromeo** [comment]
Yes I confirm I am on the master branch and I am toggling the terrains using "t" in the user interface

**Comment by orsoromeo** [comment]
This issue is probably given by a wrong installation of my Ipopt. In particular, Ipopt complains about not being able to open the `libhsl.so` library. Whenever I optimize by pressing "o" I get the following:
![ma27 solver not found](https://user-images.githubusercontent.com/14978496/39200573-d48549fc-47ec-11e8-8a98-8b3d00ec2d0c.png).

**Comment by awinkler** [comment]
It looks like you haven't installed the MA27 solver. Either change the solver as explained here: https://github.com/ethz-adrl/ifopt/issues/11
or install the solvers from here: http://www.hsl.rl.ac.uk/ipopt/
